### PR TITLE
Containerd on kubernetes 1.21+

### LINKED ISSUE #1186: Containerd add E2E tests
**What feature would you like to be added?**
At least 1 AWS test should be added (AWS to launch it every time).

A follow up based on https://github.com/kubermatic/kubeone/pull/1180#pullrequestreview-546048297

### PR DESCRIPTION
**Pull request by kron4eg**
**Which issue(s) this PR fixes** *(optional, in `fixes #<issue number>(, fixes #<issue_number>, ...)` format, will close the issue(s) when PR gets merged)*:
Fixes #1186  #1187

**Special notes for your reviewer**:
Initial test uses upgrade from 1.16 for a reason:
* to check install of old version
* to check upgrade path

**Does this PR introduce a user-facing change?**:

```release-note
NONE
```

### CODE REVIEW

**Comment by kron4eg** [comment]
/hold 

until #1204 is merged

**Comment by xmudrii** [comment]
Can you refactor this PR so it doesn’t depend on #1204? It might take some time until it’s not merged and I don’t think that we should rush it.

**Comment by kron4eg** [comment]
No, I don't think it makes sense and is not worth the effort. #1204 is logical conclusion of containerd endeavor in KubeOne and so that's why E2E is based on top of it.

**Comment by kron4eg** [comment]
Alternatively, we can close #1204 in favour of this one and close both open issues at once.

**Comment by xmudrii** [comment]
/hold

We should discuss this a bit further and double-check what's the plan for removing dockershim.

File: pkg/apis/kubeone/helpers.go
Comment (xmudrii) [inline pkg/apis/kubeone/helpers.go:94]: What is the reason for passing the byte slice as a parameter? Can we make it use string instead, especially because the function that uses this helper is using strings?

Comment (kron4eg) [reply pkg/apis/kubeone/helpers.go:94]: This is conform the https://pkg.go.dev/encoding#TextUnmarshaler interface.

File: pkg/scripts/node.go
Comment (xmudrii) [inline pkg/scripts/node.go:31]: Can we also use the Kubernetes API for draining the node, like we do for cordoning?

File: pkg/state/context.go
Comment (xmudrii) [inline pkg/state/context.go:94]: I don't really like the idea that we put this information in the state. This means that we have out-of-sync information between the state and the Cluster object. If someone would, by mistake, take `s.Cluster.ContainerRuntime`, they would get incorrect information and it would be hard to catch such a bug. It would make more sense to make this something like a defaulting function (it can't be a defaulting function, but we can re-use the mechanism), i.e. make it mutate the cluster object.

Additionally, the logic is not really correct because we don't check if the cluster is provisioned or not. I believe that this check:
```go
if condition.Check(s.LiveCluster.ExpectedVersion)
```
should become:
```go
if !s.LiveCluster.IsProvisioned() && condition.Check(s.LiveCluster.ExpectedVersion)
```

**Review by xmudrii — CHANGES_REQUESTED** [review]

File: pkg/scripts/discovery.go
Comment (kron4eg) [reply pkg/scripts/discovery.go:40]: No more of those check, the are not needed. This was a hack, and now it's time to remove it. 
Now it detects all the needed information during probes phase which became mandatory for everything.

File: pkg/scripts/node.go
Comment (kron4eg) [reply pkg/scripts/node.go:31]: Yeah, but it requires additional 200 lines of code for no good reason. Someday maybe 🤷‍♂️.

File: pkg/scripts/render.go
Comment (kron4eg) [reply pkg/scripts/render.go:129]: Well, package repo is always better over the direct download. It's checksummed, it's updating. It has dependencies that can change over time (this was the main driver to ditch the direct downloading).

After docker inc was acquired by the Mirantins, docker repositories became much better in terms of OS and versions support. And besides, we still use docker repos for docker itself. We already have the repo, so 🤷‍♂️ just use what we already have.

File: pkg/state/context.go
Comment (kron4eg) [reply pkg/state/context.go:94]: So you got me thinking again, and I think maybe we should ERROR OUT in case when the docker is specified (even of default), and plainly refuse until user will be conscious choice to change the config to indicate containerd as a container-runtime. Such approach will remove any double readings and accidents.

Comment (xmudrii) [reply pkg/state/context.go:94]: I agree, but we should start doing that only from 1.22+ because 1.21 still has Docker support. To recap, the plan should be something along:
1) Default to containerd for **new** 1.21+ clusters
2) Default to containerd for **existing** 1.22+ clusters (this should be done once we have migration in-place, not now)
3) Reject Docker for **all** clusters for Kubernetes 1.22+

Comment (kron4eg) [reply pkg/state/context.go]: OK, done, PTAL again

**Review by xmudrii — CHANGES_REQUESTED** [review]

File: pkg/state/context.go
Comment (xmudrii) [reply pkg/state/context.go]: This is still not fixed, see https://github.com/kubermatic/kubeone/pull/1205#discussion_r561732844.

File: pkg/scripts/funcs.go
Comment (kron4eg) [reply pkg/scripts/funcs.go:89]: No, luckily the CRI standard abstracts that away for us. The lowest version of kubernetes we support does support CRI (and even lesser versions). There is no such thing as "kubernetes VS containerd support matrix" as it defies the whole point of CRI.

File: pkg/apis/kubeone/helpers.go
Comment (kron4eg) [reply pkg/apis/kubeone/helpers.go:116]: I was experimenting with and without being explicit abut socket. With having it it works much better. So no, it's needed here.

File: pkg/apis/kubeone/v1beta1/defaults.go
Comment (kron4eg) [reply pkg/apis/kubeone/v1beta1/defaults.go:135]: It seems much better alternative will be:
* always default container-runtime to docker, preserving behaviour
* always require containerd to be explicitly configured as container-runtime
* fail validation for kubernetes 1.22+ if docker is indicated as container-runtime

P.S.
flatcar is working on containerd support at the moment, but it's not yet there.

Comment (kron4eg) [reply pkg/apis/kubeone/v1beta1/defaults.go:135]: The aforesaid is to avoid any "magic" jumps in logic and behaviour and keep being predictable.

Comment (xmudrii) [reply pkg/apis/kubeone/v1beta1/defaults.go:135]: The idea behind switching to containerd for the new clusters as of 1.21 is to encourage users to try and test it before we (and Kubernetes) make it mandatory as of 1.22. Of course, users could do it manually, but if we will already do it, the idea was to do it earlier (in this case 1.21).

Also, the idea was that users don't have to change their manifest (keeping backward compatibility), so if a cluster is already provisioned with Docker, we would use Docker while it's supported.

Another case that came to my mind is that if some user would stamp a custom image that has Docker preinstalled, and we default to containerd, then we might/would fail to provision the cluster. The same applies in the other direction, i.e. someone has containerd preinstalled and we try to use Docker.

I think the best scenario would be that we determine the container runtime dynamically like we determine the hostname and the operating systems. Determining container runtime would happen right after those two tasks and it would work in the following way:

* If container runtime is configured in the manifest, skip determining the CR
* Check is there Docker or containerd already installed and default to the installed CR
* If nothing is installed:
  * For Kubernetes 1.21 and newer: use containerd
  * For Kubernetes 1.20 and older: use Docker
* At the end, validate is the configuration supported (we already have that, so nothing to do from that side)

I think that this should be relatively easy to implement and would solve all our problems, while we would still be able to follow the original plan.

File: pkg/tasks/probes.go
Comment (xmudrii) [inline pkg/tasks/probes.go:133]: This logic is not going to work for newly created 1.21 clusters.

For example:
* User creates 1.21.0 cluster
* The cluster gets provisioned with containerd
* User attempts to upgrade the cluster to 1.21.1+
* This logic would default to Docker because the cluster is provisioned, so `!s.LiveCluster.IsProvisioned()` would return false
* KubeOne would install Docker and eventually break the cluster

**Review by xmudrii — CHANGES_REQUESTED** [review]

File: pkg/tasks/probes.go
Comment (kron4eg) [reply pkg/tasks/probes.go:133]: That's only seems like so, because you've missed the [static defaulting](https://github.com/kubermatic/kubeone/pull/1205/files#diff-8e3eafbecc2f7f4dbfd402187a37180eb206b8fe2128ddf3a60bb7fd2717bf07R135-R137)

Comment (kron4eg) [reply pkg/tasks/probes.go]: OK, I've misread this one. Added some live checks

**Review by xmudrii — APPROVED** [review]
/lgtm
/approve

**Pull request merged**
